### PR TITLE
Allow selecting bind address for NMI

### PR DESCRIPTION
**Pull request by whites11**
**Reason for Change**:

When using NMI on a cluster using cilium as CNI, the iptables DNAT rule that intercepts traffic towards IMDS has to be replaced by a CiliumLocalRedirectPolicy that works using eBPF.

In practice, it means that traffic is redirected to `add-pod-identity` pod's IP (which is the node IP) rather than 127.0.0.1.

This currently fails, because NMI only listens to localhost.

See https://github.com/cilium/cilium/issues/20103 for more details

This PR adds a new flag to allow specifying which address port 2579 should listen to.

**Requirements**

- [x] squashed commits
- [ ] included documentation
- [ ] added unit tests and e2e tests (if applicable). See [test standard](https://github.com/Azure/aad-pod-identity/blob/master/CONTRIBUTING.md#test-standard) for more details.
- [x] ran `make precommit`

**Issue Fixed**:

**Please answer the following questions with yes/no**:

Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

- [ ] yes
- [x] no

**Notes for Reviewers**:

The default value of the new flag is 127.0.0.1 to make this change 100% compatible with previous releases.

### CODE REVIEW

**Comment by whites11** [comment]
Tests fail because of image vulnerability, doesn't feel like related to my proposed change.

**Comment by aramase** [comment]
> Tests fail because of image vulnerability, doesn't feel like related to my proposed change.

The CVE will be patched with https://github.com/Azure/aad-pod-identity/pull/1285. Once that's merged, we'll need to rebase this PR.

File: pkg/nmi/server/server.go
Comment (aramase) [inline pkg/nmi/server/server.go:109]: The iptable rules need to use the host IP that NMI is listening on, so this needs to be updated too:
https://github.com/Azure/aad-pod-identity/blob/6cf9bac500d8f3f29492a6fd53f2e8990d51d9f0/pkg/nmi/server/server.go#L117-L119

**Review by aramase — COMMENTED** [review]
@whites11 Thanks for the PR! Added a comment. 

Our test suite will only validate with `localhost` as that's the default option. Could you confirm if you have validated a custom build with a different IP that's not localhost?

File: pkg/nmi/server/server.go
Comment (whites11) [reply pkg/nmi/server/server.go:109]: Uhm good point. Lemme think about that, it might be a little tricky

**Comment by whites11** [comment]
OK I made some changes and tested with all 3 scenarios:

Bind to all NICs:

```
$ kubectl -n giantswarm get po azure-ad-pod-identity-app-nmi-kkz8q -o yaml|grep nmi-listen
    - --nmi-listen-address=0.0.0.0

# iptables-save | grep 2579
-A aad-metadata ! -s 127.0.0.1/32 -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -j DNAT --to-destination 127.0.0.1:2579


# netstat -ntpl | grep 2579
tcp6       0      0 :::2579                 :::*                    LISTEN      504560/nmi
```

Bind to 127.0.0.1 (default)

```
$ kubectl -n giantswarm get po azure-ad-pod-identity-app-nmi-8rgvs -o yaml|grep nmi-listen
    - --nmi-listen-address=127.0.0.1
    
# iptables-save | grep 2579
-A aad-metadata ! -s 127.0.0.1/32 -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -j DNAT --to-destination 127.0.0.1:2579


# netstat -ntpl | grep 2579
tcp        0      0 127.0.0.1:2579          0.0.0.0:*               LISTEN      508739/nmi
```

Bind to a specified IP (doesn't make much sense as this IP should exist in all nodes or the pod will obviously fail to start)

```
$ kubectl -n giantswarm get po azure-ad-pod-identity-app-nmi-7zc8f -o yaml|grep nmi-listen
    - --nmi-listen-address=10.0.2.13
    
    
# iptables-save | grep 2579
-A aad-metadata ! -s 127.0.0.1/32 -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -j DNAT --to-destination 10.0.2.13:2579


# netstat -ntpl | grep 2579
tcp        0      0 10.0.2.13:2579          0.0.0.0:*               LISTEN      775371/nmi   

```

WDYT?

**Comment by aramase** [comment]
> OK I made some changes and tested with all 3 scenarios:

Yeah, that works!

**Comment by whites11** [comment]
I don't udnerstand if the failing test is a flake or some error I introduced

**Comment by whites11** [comment]
@aramase is there anything else I need to do to unblock this PR?

**Review by aramase — APPROVED** [review]

**Review by chewong — APPROVED** [review]

**Comment by whites11** [comment]
Thank you @chewong @aramase for all the help ❤️

**Comment by chewong** [comment]
@whites11 we will include this PR as part of the v1.8.10 release.

**Pull request merged**
